polaris-react/src/components/IndexTable/IndexTable.tsx

```diff
@@ -951,6 +970,7 @@ function IndexTableBase({
             heading?.alignment === 'end' &&
             styles['TableHeadingSortButton-heading-align-end-currently-sorted'],
           isPreviouslySorted &&
+            !isRenderAfterSelectEvent &&
```

Comment:
This prevents the sort button from sliding after a right-aligned header is selected, then a different sort header is clicked, and then a single row checkbox or the "select all" checkbox is clicked.